#!/usr/bin/bash


#BAMLIST=$1   # List of bam file samples
#GENOMEREF=$2 # Genome reference FASTA file
#GENOMEFAI=$3 # Genome FAI generated by samtools
#CHUNKS=$4    # Length of the chunks to process in parallel
#OUTDIR=$5    # Output directory where to store intermediate and final SNP calls

BAMLIST=$PWD/SBS_population.txt
GENOMEREF=/users/fk/mkonczal/projects/SBS/assembly/final/SBS_final.scaffolds.fasta
GENOMEFAI=/users/fk/mkonczal/projects/SBS/assembly/final/SBS_final.scaffolds.fasta.fai
CHUNKS=1000000
OUTDIR=$PWD/ChunksSBS

mkdir -p $OUTDIR

# Get the chromosome split file
vcfutils.pl splitchr -l $CHUNKS $GENOMEFAI > $OUTDIR/splitchr.txt
# Number of jobs for jarray from chromosome split file

njobs=$(wc -l $OUTDIR/splitchr.txt | awk '{print $1}')
##njobs=5

# Send all the splits for snp calling in parallel

jobid=`qsub -t 1-$njobs -N SNPcalling -wd $OUTDIR -l virtual_free=6G,h_rt=06:00:00 SNP_calling_chunks.sh $BAMLIST $GENOMEREF splitchr.txt`
